pulsar/internal/connection.go

```diff
@@ -331,10 +335,15 @@ func (c *connection) waitUntilReady() error {
 }
 
 func (c *connection) failLeftRequestsWhenClose() {
+	// wait for outstanding incoming requests to complete before draining
+	// and closing the channel
+	c.incomingRequestsWG.Wait()
```

Comment:
I'm a little confused why are we waiting here?

Reply:
@cckellogg ,
At this point we are sure that the closeCh is closed and that the state of the connection is either connectionClosing or connectionClosed.  We wait to be sure that there will be no further writes to the incomingRequestsCh before attempting to drain and close it.  It is possible for the sending go-routine to be executing either of the following lines at the time this function is executed:
- https://github.com/apache/pulsar-client-go/blob/master/pulsar/internal/connection.go#L552
- https://github.com/apache/pulsar-client-go/blob/master/pulsar/internal/connection.go#L570 

As mentioned in the motivation for this MR:

> When two cases of a select are valid, the case executed is chosen at random; see https://tour.golang.org/concurrency/5

Whenever I see a channel being closed by a reader go-routine I am suspicious of writers attempting to write to the channel after it is closed.